### ISSUE TITLE
ENAS-micro_search-performance

### ISSUE THREAD

**Issue by zxlgit123** [microsoft/nni#3447]
## NNI 学生项目问题概述 / General Question of Student Program

**请简要概述您的问题 / 观点 :**
**Short summary about the question / idea :**
我用nni中enas的micro默认参数运行cifar10的数据，性能只有46%多点
参数信息详细为：
micro，cifar10，epoch=200，batch_size=32，num_layers=6，out_channels=20，num_nodes=5，dropout_rate=0.1，use_aux_heads=True，tanh_constant=1.1，cell_exit_extra_step=True

执行代码路径是：python /nni/examples/nas/enas/search.py

**我想问下，我们nni自己测试nni这个参数大概的最高精度是多少呀？特别期待您的回复，谢谢！**

**请提供 NNI 环境信息 :**
**nni Environment :**  pytorch nvidia-v100
- nni version: v1.8
- nni mode(local|pai|remote): local
- OS: Linux
- python version: 3.7
- is conda or virtualenv used?: conda
- is running in docker?: yes

## 其他建议 / Other Advice

**是否需要更新文档（是 / 否）:**
**Need to update document ( yes / no ) :**

**其他分享内容 :**
**Anything else we need to know :**

**Log message / 日志信息 :**
 - [nnimanager.log and dispatcher.log](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/HowToDebug.md#experiment-root-directory) : 

 - [nnictl stdout and stderr](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/Nnictl.md#nnictl%20log%20stdout) :

**Comment by ultmaster** [comment]
你 search 出来的只是 architecture，architecture 需要重新训练才行。重新训练的 code 可以参考 DARTS。

**Comment by zxlgit123** [comment]
> 你 search 出来的只是 architecture，architecture 需要重新训练才行。重新训练的 code 可以参考 DARTS。

好的，我看下。谢谢啦。

**Comment by zxlgit123** [comment]
> > 你 search 出来的只是 architecture，architecture 需要重新训练才行。重新训练的 code 可以参考 DARTS。
> 
> 好的，我看下。谢谢啦。

您好，我重新跑了一下参数改为：batch_size=16，num_layers=12，out_channels=36，num_nodes=6，跑出来的结果是60.33%。加了训练的结果是66.73%，这个精度是合理的吗？epoch=50。您那边的精度在cifar10上都大概是什么样子的呢？

**Comment by ultmaster** [comment]
你先试试 darts 官方搜出来的 architecture 能 train 出多少。。cifar10 起码要 97 才算合理吧

**Comment by zxlgit123** [comment]
> 你先试试 darts 官方搜出来的 architecture 能 train 出多少。。cifar10 起码要 97 才算合理吧

我们现在主要想用enas的微观搜索加训练。您那边如果只有enas的话跑出来的cifar10最高精度是多少啊？加了训练之后是多少呢？
我在nni默认参数下，这么低的精度您能帮忙分析一下是哪方面的原因呢？

**Comment by ultmaster** [comment]
你能 contribute 一个 enas 的 retrain 的 code 吗？我可以给你 review 一下

**Comment by zxlgit123** [comment]
> 你能 contribute 一个 enas 的 retrain 的 code 吗？我可以给你 review 一下

谢谢您，下面是我的训练代码，您帮我看看。
def accuracy(output, target, topk=(1,)):
    with torch.no_grad():
        maxk = max(topk)
        batch_size = target.size(0)

        _, pred = output.topk(maxk, 1, True, True)
        pred = pred.t()
        correct = pred.eq(target.view(1, -1).expand_as(pred))

        res = []
        for k in topk:
            correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
            res.append(correct_k.mul_(100.0 / batch_size))
        return res
class AverageMeter(object):
    def __init__(self, name, fmt=':f'):
        self.name = name
        self.fmt = fmt
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

    def __str__(self):
        fmtstr = '{name} {val' + self.fmt + '} ({avg' + self.fmt + '})'
        return fmtstr.format(**self.__dict__)


def train(train_loader, model, criterion, optimizer, epoch, args, device, logger):
    losses = AverageMeter('Loss', ':.4e')
    top1 = AverageMeter('Acc@1', ':6.2f')
    top5 = AverageMeter('Acc@5', ':6.2f')

    # 模型训练
    model.train()
    logger.info("Epoch %d starts" % epoch)

    for i, (images, target) in enumerate(train_loader):
        images =images.to(device)
        target = target.to(device)
        # compute output
        output = model(images)

        loss = criterion(output[0], target)
        acc1, acc5 = accuracy(output[0], target, topk=(1, 5))
        
        losses.update(loss.item(), images.size(0))
        top1.update(acc1[0], images.size(0))
        top5.update(acc5[0], images.size(0))


        # compute gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        current_step = len(train_loader) * epoch + i

        if i % 10 == 0:
            logger.info(f"train/loss is {losses.val}/{current_step}")
            logger.info(f"train/acc_1 is {top1.val}/{current_step}")
            logger.info(f"train/acc_1 is {top5.val}/{current_step}")
            
    return model,top1,True,losses

**Comment by zxlgit123** [comment]
> 你能 contribute 一个 enas 的 retrain 的 code 吗？我可以给你 review 一下

`def accuracy(output, target, topk=(1,)):
    with torch.no_grad():
        maxk = max(topk)
        batch_size = target.size(0)

        _, pred = output.topk(maxk, 1, True, True)
        pred = pred.t()
        correct = pred.eq(target.view(1, -1).expand_as(pred))

        res = []
        for k in topk:
            correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
            res.append(correct_k.mul_(100.0 / batch_size))
        return res
class AverageMeter(object):
    """Computes and stores the average and current value"""

    def __init__(self, name, fmt=':f'):
        self.name = name
        self.fmt = fmt
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

    def __str__(self):
        fmtstr = '{name} {val' + self.fmt + '} ({avg' + self.fmt + '})'
        return fmtstr.format(**self.__dict__)


def train(train_loader, model, criterion, optimizer, epoch, args, device, logger):
    losses = AverageMeter('Loss', ':.4e')
    top1 = AverageMeter('Acc@1', ':6.2f')
    top5 = AverageMeter('Acc@5', ':6.2f')

    # 模型训练
    model.train()
    logger.info("Epoch %d starts" % epoch)

    for i, (images, target) in enumerate(train_loader):
        images =images.to(device)
        target = target.to(device)
        # compute output
        output = model(images)

        loss = criterion(output[0], target)
        acc1, acc5 = accuracy(output[0], target, topk=(1, 5))
        
        losses.update(loss.item(), images.size(0))
        top1.update(acc1[0], images.size(0))
        top5.update(acc5[0], images.size(0))


        # compute gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        current_step = len(train_loader) * epoch + i

        if i % 10 == 0:
            logger.info(f"train/loss is {losses.val}/{current_step}")
            logger.info(f"train/acc_1 is {top1.val}/{current_step}")
            logger.info(f"train/acc_1 is {top5.val}/{current_step}")
            
    return model,top1,True,losses`

**Comment by zxlgit123** [comment]
> 你能 contribute 一个 enas 的 retrain 的 code 吗？我可以给你 review 一下

下面是我的代码，您帮我看看哪里不太对，谢谢啦。

def accuracy(output, target, topk=(1,)):
    with torch.no_grad():
        maxk = max(topk)
        batch_size = target.size(0)

        _, pred = output.topk(maxk, 1, True, True)
        pred = pred.t()
        correct = pred.eq(target.view(1, -1).expand_as(pred))

        res = []
        for k in topk:
            correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
            res.append(correct_k.mul_(100.0 / batch_size))
        return res
class AverageMeter(object):
    """Computes and stores the average and current value"""

    def __init__(self, name, fmt=':f'):
        self.name = name
        self.fmt = fmt
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

    def __str__(self):
        fmtstr = '{name} {val' + self.fmt + '} ({avg' + self.fmt + '})'
        return fmtstr.format(**self.__dict__)


def train(train_loader, model, criterion, optimizer, epoch, args, device, logger):
    losses = AverageMeter('Loss', ':.4e')
    top1 = AverageMeter('Acc@1', ':6.2f')
    top5 = AverageMeter('Acc@5', ':6.2f')

    # 模型训练
    model.train()
    logger.info("Epoch %d starts" % epoch)

    for i, (images, target) in enumerate(train_loader):
        images =images.to(device)
        target = target.to(device)
        # compute output
        output = model(images)

        loss = criterion(output[0], target)
        acc1, acc5 = accuracy(output[0], target, topk=(1, 5))
        
        losses.update(loss.item(), images.size(0))
        top1.update(acc1[0], images.size(0))
        top5.update(acc5[0], images.size(0))


        # compute gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        current_step = len(train_loader) * epoch + i

        if i % 10 == 0:
            logger.info(f"train/loss is {losses.val}/{current_step}")
            logger.info(f"train/acc_1 is {top1.val}/{current_step}")
            logger.info(f"train/acc_1 is {top5.val}/{current_step}")
            
    return model,top1,True,losses

**Comment by zxlgit123** [comment]
> 你能 contribute 一个 enas 的 retrain 的 code 吗？我可以给你 review 一下

您好，您帮我看了代码有啥问题么？

**Comment by ultmaster** [comment]
我的意思是你开个 PR 可以吗？

**Comment by zxlgit123** [comment]
> PR

我还不太会用github，哎，伤心，我学习一下再看看吧。
